mm2src/kdf_walletconnect/src/client/stream.rs

```diff
@@ -0,0 +1,293 @@
+}
+
+impl ClientStream {
+    fn new(socket: WebSocketStream) -> Self {
+        let id_generator = MessageIdGenerator::new();
+        let (outbound_tx, outbound_rx) = unbounded_channel();
+        let requests = HashMap::new();
+
+        Self {
+            socket,
+            outbound_rx,
+            outbound_tx,
+            requests,
+            close_frame: None,
+            id_generator,
+            socket_ended: false,
+        }
+    }
+
+    /// Sends an already serialized [`OutboundRequest`][OutboundRequest] (see
+    /// [`create_request()`]).
+    pub fn send_raw(&mut self, request: OutboundRequest) {
+        let tx = request.tx;
+        let id = self.id_generator.next();
+        let request = Payload::Request(rpc::Request::new(id, request.params));
+        let serialized = serde_json::to_string(&request);
+
+        match serialized {
+            Ok(data) => match self.requests.entry(id) {
+                Entry::Occupied(_) => {
+                    tx.send(Err(ClientError::DuplicateRequestId)).ok();
+                },
+
+                Entry::Vacant(entry) => {
+                    entry.insert(tx);
+                    self.outbound_tx.send(Message::Text(data)).ok();
+                },
+            },
+
+            Err(err) => {
+                tx.send(Err(ClientError::SerdeError(err.to_string()))).ok();
+            },
+        }
+    }
+
+    /// Serialize the request into a generic [`OutboundRequest`] and sends it,
+    /// returning a future that resolves with the response.
+    pub fn send<T>(&mut self, request: T) -> ResponseFuture<T>
+    where
+        T: ServiceRequest,
+    {
+        let (request, response) = create_request(request);
+        self.send_raw(request);
+        response
+    }
+
+    /// Closes the connection.
+    pub async fn close(&mut self) -> MmResult<(), ClientError> {
+        self.socket
+            .close()
+            .await
+            .map_to_mm(|err| WebsocketClientError::ClosingFailed(format!("{err:?}")).into())
+    }
+
+    fn parse_inbound(&mut self, result: Result<Message, WebsocketClientError>) -> Option<StreamEvent> {
+        match result {
+            Ok(message) => match &message {
+                Message::Binary(_) | Message::Text(_) => {
+                    let payload: Payload = match serde_json::from_slice(&message.into_data()) {
+                        Ok(payload) => payload,
+
+                        Err(err) => return Some(StreamEvent::InboundError(ClientError::SerdeError(err.to_string()))),
+                    };
+
+                    match payload {
+                        Payload::Request(request) => {
+                            let id = request.id;
+
+                            let event = match request.params {
+                                Params::Subscription(data) => StreamEvent::InboundSubscriptionRequest(
+                                    InboundRequest::new(id, data, self.outbound_tx.clone()),
+                                ),
```

Comment:
I can't really follow why we use an unbounded response channel here? 
isn't it gonne be just one resposne? it also seems not to be used anywhere later in `PublishedMessage`.

Reply:
note: This code is part of original code from WalletConnect, I also had similar question but I think there might be a good reason to use `unbounded_channel` e.g when we don't want to miss a message or have sender blocking